data race in TestUpdateNodeStatusWithLease

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93605/pull-kubernetes-bazel-test/1290329488629960704
```
=== RUN   TestUpdateNodeStatusWithLease
I0803 17:08:37.155082      23 plugin_manager.go:114] Starting Kubelet Plugin Manager
==================
WARNING: DATA RACE
Read at 0x00c00066d938 by goroutine 106:
  k8s.io/kubernetes/pkg/kubelet/cadvisor.CapacityFromMachineInfo()
      pkg/kubelet/cadvisor/util.go:44 +0x1d5
  k8s.io/kubernetes/pkg/kubelet/nodestatus.MachineInfo.func1()
      pkg/kubelet/nodestatus/setters.go:287 +0x19dc
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).setNodeStatus()
      pkg/kubelet/kubelet_node_status.go:559 +0x154
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initialNode()
      pkg/kubelet/kubelet_node_status.go:409 +0xdcf
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).GetNode()
      pkg/kubelet/kubelet_getters.go:236 +0x14a
  k8s.io/kubernetes/pkg/kubelet/server/stats.(*summaryProviderImpl).Get()
      pkg/kubelet/server/stats/summary.go:70 +0x7e
  k8s.io/kubernetes/pkg/kubelet/server/stats.(*resourceAnalyzer).Get()
      <autogenerated>:1 +0x74
  k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).synchronize()
      pkg/kubelet/eviction/eviction_manager.go:258 +0x14b
  k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).Start.func2()
      pkg/kubelet/eviction/eviction_manager.go:203 +0x66

Previous write at 0x00c00066d938 by goroutine 104:
  [failed to restore the stack]

Goroutine 106 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/eviction.(*managerImpl).Start()
      pkg/kubelet/eviction/eviction_manager.go:201 +0x304
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules()
      pkg/kubelet/kubelet.go:1296 +0x656
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules-fm()
      pkg/kubelet/kubelet.go:1274 +0x44
  sync.(*Once).doSlow()
      GOROOT/src/sync/once.go:66 +0x109
  sync.(*Once).Do()
      GOROOT/src/sync/once.go:57 +0x68
  k8s.io/kubernetes/pkg/kubelet.(*Kubelet).updateRuntimeUp()
      pkg/kubelet/kubelet.go:2116 +0x652
  k8s.io/kubernetes/pkg/kubelet.TestUpdateNodeStatusWithLease()
      pkg/kubelet/kubelet_node_status_test.go:904 +0x2739
  testing.tRunner()
      GOROOT/src/testing/testing.go:1108 +0x202

Goroutine 104 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1159 +0x796
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1430 +0xa6
  testing.tRunner()
      GOROOT/src/testing/testing.go:1108 +0x202
  testing.runTests()
      GOROOT/src/testing/testing.go:1428 +0x5aa
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1338 +0x4eb
  main.main()
      bazel-out/k8-fastbuild/bin/pkg/kubelet/linux_amd64_race_stripped/go_default_test%/testmain.go:266 +0x304
  runtime.main()
      GOROOT/src/runtime/proc.go:204 +0x208
  k8s.io/kubernetes/vendor/github.com/cilium/ebpf.init()
      vendor/github.com/cilium/ebpf/prog.go:362 +0x3f8
  k8s.io/kubernetes/vendor/github.com/cilium/ebpf.init()
      vendor/github.com/cilium/ebpf/syscalls.go:392 +0x38b
  k8s.io/kubernetes/vendor/github.com/cilium/ebpf.init()
      vendor/github.com/cilium/ebpf/syscalls.go:216 +0x31d
  k8s.io/kubernetes/vendor/github.com/cilium/ebpf.init()
      vendor/github.com/cilium/ebpf/syscalls.go:188 +0x2b0
  runtime.doInit()
      GOROOT/src/runtime/proc.go:5590 +0x89
  k8s.io/kubernetes/vendor/github.com/cilium/ebpf/internal/btf.init()
      vendor/github.com/cilium/ebpf/internal/btf/btf.go:656 +0x18f
==================
E0803 17:08:37.174713      23 eviction_manager.go:260] eviction manager: failed to get summary stats: failed to get root cgroup stats: failed to get cgroup stats for "/": unexpected number of containers: 0
I0803 17:08:37.178686      23 kubelet_network.go:77] Setting Pod CIDR:  -> 10.0.0.0/24,2000::/10
    testing.go:1023: race detected during execution of test
--- FAIL: TestUpdateNodeStatusWithLease (0.04s)
```

the test sets `kubelet.machineInfo.MemoryCapacity` in an unprotected way that races with the background eviction manager goroutine, which calls GetCachedMachineInfo, also in an unprotected way

_Originally posted by @liggitt in https://github.com/kubernetes/kubernetes/pull/93605#issuecomment-668141263_

/sig node
/kind flake
/cc @dashpole

--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -561,7 +561,7 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	if err != nil {
 		return nil, err
 	}
-	klet.machineInfo = machineInfo
+	klet.setCachedMachineInfo(machineInfo)
 
 	imageBackOff := flowcontrol.NewBackOff(backOffPeriod, MaxContainerBackOff)
 
@@ -911,7 +911,8 @@ type Kubelet struct {
 	configMapManager configmap.Manager
 
 	// Cached MachineInfo returned by cadvisor.
-	machineInfo *cadvisorapi.MachineInfo
+	machineInfoLock sync.RWMutex
+	machineInfo     *cadvisorapi.MachineInfo
 
 	// Handles certificate rotations.
 	serverCertificateManager certificate.Manager

--- a/pkg/kubelet/kubelet_getters.go
+++ b/pkg/kubelet/kubelet_getters.go
@@ -382,5 +382,13 @@ func (kl *Kubelet) GetVersionInfo() (*cadvisorapiv1.VersionInfo, error) {
 
 // GetCachedMachineInfo assumes that the machine info can't change without a reboot
 func (kl *Kubelet) GetCachedMachineInfo() (*cadvisorapiv1.MachineInfo, error) {
+	kl.machineInfoLock.RLock()
+	defer kl.machineInfoLock.RUnlock()
 	return kl.machineInfo, nil
+}
+
+func (kl *Kubelet) setCachedMachineInfo(info *cadvisorapiv1.MachineInfo) {
+	kl.machineInfoLock.Lock()
+	defer kl.machineInfoLock.Unlock()
+	kl.machineInfo = info
 }

--- a/pkg/kubelet/kubelet_node_status_test.go
+++ b/pkg/kubelet/kubelet_node_status_test.go
@@ -212,7 +212,7 @@ func TestUpdateNewNodeStatus(t *testing.T) {
 				NumCores:       2,
 				MemoryCapacity: 10e9, // 10G
 			}
-			kubelet.machineInfo = machineInfo
+			kubelet.setCachedMachineInfo(machineInfo)
 
 			expectedNode := &v1.Node{
 				ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname},
@@ -390,7 +390,7 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 20e9,
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	expectedNode := &v1.Node{
 		ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname},
@@ -596,7 +596,7 @@ func TestUpdateNodeStatusWithRuntimeStateError(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 10e9,
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	expectedNode := &v1.Node{
 		ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname},
@@ -816,7 +816,7 @@ func TestUpdateNodeStatusWithLease(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 20e9,
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	now := metav1.NewTime(clock.Now()).Rfc3339Copy()
 	expectedNode := &v1.Node{
@@ -961,7 +961,13 @@ func TestUpdateNodeStatusWithLease(t *testing.T) {
 	// Report node status even if it is still within the duration of nodeStatusReportFrequency.
 	clock.Step(10 * time.Second)
 	var newMemoryCapacity int64 = 40e9
-	kubelet.machineInfo.MemoryCapacity = uint64(newMemoryCapacity)
+	oldMachineInfo, err := kubelet.GetCachedMachineInfo()
+	if err != nil {
+		t.Fatal(err)
+	}
+	newMachineInfo := oldMachineInfo.Clone()
+	newMachineInfo.MemoryCapacity = uint64(newMemoryCapacity)
+	kubelet.setCachedMachineInfo(newMachineInfo)
 	assert.NoError(t, kubelet.updateNodeStatus())
 
 	// 2 more action (There were 5 actions before).
@@ -1073,7 +1079,7 @@ func TestUpdateNodeStatusAndVolumesInUseWithNodeLease(t *testing.T) {
 			kubelet.containerManager = &localCM{ContainerManager: cm.NewStubContainerManager()}
 			kubelet.lastStatusReportTime = kubelet.clock.Now()
 			kubelet.nodeStatusReportFrequency = time.Hour
-			kubelet.machineInfo = &cadvisorapi.MachineInfo{}
+			kubelet.setCachedMachineInfo(&cadvisorapi.MachineInfo{})
 
 			// override test volumeManager
 			fakeVolumeManager := kubeletvolume.NewFakeVolumeManager(tc.existingVolumes)
@@ -1147,7 +1153,7 @@ func TestRegisterWithApiServer(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 1024,
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	done := make(chan struct{})
 	go func() {
@@ -1356,7 +1362,7 @@ func TestUpdateNewNodeStatusTooLargeReservation(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 10e9, // 10G
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	expectedNode := &v1.Node{
 		ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname},
@@ -2183,7 +2189,7 @@ func TestRegisterWithApiServerWithTaint(t *testing.T) {
 		NumCores:       2,
 		MemoryCapacity: 1024,
 	}
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	var gotNode runtime.Object
 	kubeClient.AddReactor("create", "nodes", func(action core.Action) (bool, runtime.Object, error) {

--- a/pkg/kubelet/kubelet_test.go
+++ b/pkg/kubelet/kubelet_test.go
@@ -215,7 +215,7 @@ func newTestKubeletWithImageList(
 
 	kubelet.cadvisor = &cadvisortest.Fake{}
 	machineInfo, _ := kubelet.cadvisor.MachineInfo()
-	kubelet.machineInfo = machineInfo
+	kubelet.setCachedMachineInfo(machineInfo)
 
 	fakeMirrorClient := podtest.NewFakeMirrorClient()
 	secretManager := secret.NewSimpleSecretManager(kubelet.kubeClient)

